### ISSUE TITLE
Multiplying by `cos` angle correction factors results in slightly different values (~1e-3) compared to `-angle-corr 1`

### ISSUE THREAD

**Issue by joshuacwnewton** [spinalcordtoolbox/spinalcordtoolbox#4378]
While investigating #4367, I tried validating the suggestion by Julien (https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4367#issuecomment-1948521324) by trying to apply angle correction 2 different ways (on just the spinal cord segmentation, i.e. no GM/WM involved in scope of this issue):

1. Automatically, using SCT's usual `-angle-corr 1` behavior.
2. Manually, by following @jcohenadad's [suggestion](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4367#issuecomment-1948521324) of multiplying by `cos(angle_RP/180*PI())` and `cos(angle_AP/180*PI())` using Google Sheets' `COS` and `PI` functions.

Green is the output of `-angle-corr 0`, Orange is manual multiplication of angle values, Blue is `-angle-corr 1`:

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/4fa05bbf-4d80-47dd-824c-918630b4c504)

Both the orange and blue values ~_should_ be the same (since we're just multiplying the uncorrected area by `cos(angle)*cos(angle)` in both cases)~. However, they are not equal.

I will try to investigate why this happens and make sure we are outputting the "correct" angle-corrected values.

**Comment by joshuacwnewton** [comment]
> * The angle values written to the spreadsheet are different than the values used in-code, so my manual multiplication is "off" compared to the computed values in the Python script.
> * The conversion from degrees to radians somehow introduces a loss of precision.

Here are the angle values (and cos values) used in code for slice `0`:

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/5f4c84f1-38ec-4b79-8324-43794c2cbbf4)

Comparing above, we can see that these values are identical to those computed in row 1 of the  spreadsheet. 

----

So, I can only conclude that the source of the error stems from: 

> * Differences between spreadsheet's multiplication vs. skimage.transform.AffineTransform (which is used to scale the values in the Python script).

For Google Sheets, I simply multiplied the uncorrected area by the 2 cosine values (`=B2*D2*F2`). 

For the Python code, the metrics are computed via these calls:

```python
tform = transform.AffineTransform(scale=(angle_RL_cos, angle_AP_cos))
current_patch_scaled = transform.warp(current_patch,
                                      tform.inverse,
                                      output_shape=current_patch.shape,
                                      order=1,
                                                  )
shape_property = _properties2d(current_patch_scaled, [px, py])  # Contains `'area'` key
```

So, something in the above code is causing some sort of precision loss.

**Comment by joshuacwnewton** [comment]
Starting with `_properties2d`, the area is computed using this logic:

```python
    upscale = 5  # upscale factor for resampling the input image (for better precision)

    # Oversample image to reach sufficient precision when computing shape metrics on the binary mask
    image_crop_r = transform.pyramid_expand(image_crop, upscale=upscale, sigma=None, order=1)

    # Compute area with weighted segmentation and adjust area with physical pixel size
    area = np.sum(image_crop_r) * dim[0] * dim[1] / upscale ** 2
```

For `-angle-corr 0`:

Calculation                              | Value
---------------------------------------- | ------
`np.sum(image_crop_r)`                   | 25000.00
`dim[0]`/`dim[1]`                        | 0.05
`np.sum(image_crop_r) * dim[0] * dim[1]` | 62.5
` / upscale ** 2`                        | 2.5
`np.sum(image_crop_r) * dim[0] * dim[1] / upscale ** 2` | 2.5000000745058064

For `-angle-corr 1`:

Calculation            | Value
---------------------- | ------
`np.sum(image_crop_r)` | 22107.686376004716
` / upscale ** 2`      | 2.210768703486512

Given the fact that, in this specific case, the upscaling and the resolution of the image (5, 0.05) line up nicely, it at least demonstrates to me that our area computation steps are relatively simple/harmless. 

(I'm noticing some slight precision errors in the full area computation, which is curious! Though, I don't think this explains the errors of magnitude `1e-3`. Still, I'm going to try correct this first before checking anything else -- the output of the uncorrected SC should read 2.5, not 2.5000000745058064!)

**Comment by joshuacwnewton** [comment]
Some testing:

```python
>>> 25000.0 * 0.05 * 0.05
62.5
>>> area_pix = 25000.0
>>> dim = 0.05
>>> area_pix * dim * dim
62.5
>>> import numpy as np
>>> area_pix_np = np.float64(25000.0)
>>> dim_np = np.float64(0.05)
>>> area_pix_np * dim_np * dim_np
62.5
```

```python
>>> np.sum(image_crop_r)
25000.0
>>> dim[0]
0.05
>>> dim[1]
0.05
dim[0] * dim[1]
0.0025000002
np.sum(image_crop_r) * dim[0]
1250.0000186264515
>>> np.sum(image_crop_r) * dim[0] * dim[1]
62.50000186264516
```

I'm wondering if the floating point issues witnessed above are at least in part due to https://docs.python.org/3/tutorial/floatingpoint.html.

> Many users are not aware of the approximation because of the way values are displayed. Python only prints a decimal approximation to the true decimal value of the binary approximation stored by the machine. On most machines, if Python were to print the true decimal value of the binary approximation stored for 0.1, it would have to display:
> ```python
> >>> 0.1
> 0.1000000000000000055511151231257827021181583404541015625
> ```

That page mentions `.as_integer_ratio`. Applying it to the values above:

```python
>>> np.sum(image_crop_r).as_integer_ratio()
(25000, 1)
>>> dim[0].as_integer_ratio()
(13421773, 268435456)
>>> 0.5.as_integer_ratio()
(1, 2)
```

So, that at least partially explains things! The dim values themselves are lacking even more precision than Python's built-in floats. 

For the sake of testing only, I'm going to round these `dim` values and continue investigating.

**Comment by joshuacwnewton** [comment]
Testing `skimage.transform.AffineTransform`: 

```python
# compute non-angle-corrected shape properties
current_patch = current_patch.astype(np.float64)
shape_property_unscaled = _properties2d(current_patch, [px, py])

# compute angle corrected shape properties
tform = transform.AffineTransform(scale=(angle_RL_cos, angle_AP_cos))
current_patch_scaled = transform.warp(current_patch,
                                      tform.inverse,
                                      output_shape=current_patch.shape,
                                      order=1,
                                      )
shape_property_scaled = _properties2d(current_patch_scaled, [px, py])

# compare skimage-scaled area to manually scaled area
>>> print(shape_property_unscaled['area'])
2.5
>>> print(shape_property_unscaled['area'] * angle_RL_cos * angle_AP_cos)
2.2099990766318953
>>> print(shape_property_scaled['area'])
2.2107686375537283
```

Manually inspecting the arrays `current_patch` and `current_patch_scaled`, I see:

<details>

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/57c586ce-5976-4431-bf8a-e364f08887af)

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/83f4c439-a072-4759-8463-865fcc038480)

</details>

Seeing this visualized helped me to understand why there might be a difference:

- Approach 1: Google Sheets
  - Compute voxelwise area of binary (`1`/`0`) mask by summing 2D pixels, then multiply by `cos(angle_AP)*cos(angle_RL)`
  - This method essentially treats the binary mask's spinal cord area as equivalent to a rectangle, and the area as HxW -- we scale the sides of the rectangle using the `cos` values, no interpolation used.
- Approach 2: Scale the cord as an affine transformation (with interpolation), _then_ sum the values.
  - This necessarily requires interpolation at the interface, as displayed in the screenshots above.

So, I _think_ the differences can be explained away simply as interpolation error? Since the cord isn't a perfect square, the `skimage`-based scaling is going to be necessarily more complex than just multiplying the edges of an equivalent-area rectangle by the scaling values.

To demonstrate, I iterated through rectangles with the same pixelwise area (1000) as the slice above:

<details>

```python
            for factor in (8, 10, 20, 25, 40, 50, 100, 125):
                factor2 = 1000//factor
                midpoint = 100
                start = (midpoint - (factor//2), midpoint - (factor2//2))
                extent = (factor, factor2)
                from skimage.draw import rectangle
                patch = np.zeros((200, 200), dtype=np.uint8)
                rr, cc = rectangle(start, extent=extent, shape=patch.shape)
                patch[rr, cc] = 1
                patch_scaled = transform.warp(patch, tform.inverse, output_shape=patch.shape, order=1)
                shape_property_unscaled = _properties2d(patch, [px, py])
                shape_property_scaled = _properties2d(patch_scaled, [px, py])
                print(f"Mask shape: {(factor, factor2)} | "
                      f"Uncorrected area: {shape_property_unscaled['area']} | "
                      f"Corrected area (mult): {shape_property_unscaled['area'] * angle_AP_cos * angle_RL_cos} | "
                      f"Corrected area (interp): {shape_property_scaled['area']}")
```

</details>

Result:

```python
Mask shape: (8, 125)  | Uncorrected area: 2.5 | Corrected area (mult): 2.2099990766318953 | Corrected area (interp): 2.2181110036887515
Mask shape: (10, 100) | Uncorrected area: 2.5 | Corrected area (mult): 2.2099990766318953 | Corrected area (interp): 2.212536911441094
Mask shape: (20, 50)  | Uncorrected area: 2.5 | Corrected area (mult): 2.2099990766318953 | Corrected area (interp): 2.2020101194973045
Mask shape: (25, 40)  | Uncorrected area: 2.5 | Corrected area (mult): 2.2099990766318953 | Corrected area (interp): 2.2169827804541473
Mask shape: (40, 25)  | Uncorrected area: 2.5 | Corrected area (mult): 2.2099990766318953 | Corrected area (interp): 2.207230914894842
Mask shape: (50, 20)  | Uncorrected area: 2.5 | Corrected area (mult): 2.2099990766318953 | Corrected area (interp): 2.2138920446967036
Mask shape: (100, 10) | Uncorrected area: 2.5 | Corrected area (mult): 2.2099990766318953 | Corrected area (interp): 2.2148820561415246
Mask shape: (125, 8)  | Uncorrected area: 2.5 | Corrected area (mult): 2.2099990766318953 | Corrected area (interp): 2.208714256522523
```

In all cases, the "simple multiplication" area is the same as the SC cord from this issue (2.209999). But, the interpolated area varies wildly depending on both the shape and orientation of the rectangle. 

In other words -- Does this suggest that the "simple multiplication" approach is inaccurate? Or does it suggest that there's a limit to the precision of the area we can compute using a `0.05x0.05mm` voxel resolution in the first place?

**Comment by joshuacwnewton** [comment]
This is probably all very naive of me, to be quite honest. :sweat_smile: 

In hindsight, I'm wondering why we don't repeatedly apply a rotational affine transformation to the cord as a whole (using this same `skimage` method), then compute the cross-sectional area for the plane orthogonal to the cord at each point in the centerline. Similar to what's visualized here:

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/9dc988cc-1acd-41b5-96ee-5f2c8922f4a1)

Why do we instead take only axial slices? Surely that's less accurate at capturing the cross-section of the cord, right? (Especially for something like GM/WM?)

**Comment by valosekj** [comment]
Super interesting investigation! 💯 

> This method essentially treats the binary mask's spinal cord area as equivalent to a rectangle, and the area as HxW -- we scale the sides of the rectangle using the cos values, no interpolation used.

What will be the effect of treating the SC as a rectangle when working, for example, with a compressed cord?

<img width="555" alt="image" src="https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/39456460/b19ab527-e95e-4ab9-8371-ae278c5da8d4">

**Comment by joshuacwnewton** [comment]
> What will be the effect of treating the SC as a rectangle when working, for example, with a compressed cord?

Just for clarification, "treating the SC as a rectangle" was just a temporary workaround suggested by Julien [here](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4367#issuecomment-1948521324), to avoid having to implement a new command line flag.

(So, in essence, this entire issue is only relevant for that comment specifically, and not necessarily relevant for `sct_process_segmentation` as a whole.)

But, to answer your question -- I don't think compression changes anything? My observations were basically:

- If we use `-angle-corr 1`, the shape of the cord will cause variations in the total area due to interpolation.
- If we simply multiply the "uncorrected" binary area by the cos factors, the resulting "corrected" area will be the same regardless of the shape.

**Comment by valosekj** [comment]
Okay, I see! Thank you for the clarification!

**Comment by joshuacwnewton** [comment]
As discussed in our 2024-02-21 meeting (as well as privately via email with @BFCombes), we've determined that these small differences are not critical, and do not represent a bug or issue in SCT.

However, I have some thoughts as to how we could avoid this confusion in the future:

- Perhaps we could update the documentation to explain that "There is a limit to the level of precision you can achieve for a given resolution"
- Perhaps we could even intentionally truncate the [non-significant figures](https://en.wikipedia.org/wiki/Significant_figures) that arise from the angle correction:
    > The following types of digits are not considered significant:[[2]](https://en.wikipedia.org/wiki/Significant_figures#cite_note-2)
    > - [Spurious](https://en.wiktionary.org/wiki/spurious) digits that arise from calculations resulting in a higher precision than the original data or a measurement reported with greater precision than the instrument's resolution.

  Perhaps if we were to do this, then both angle correction methods _would_ end up with the same results. :thinking: 

  Given that we have both the # of binary voxels + the physical dimensions of a voxel (pixdim), it should be easy to compute the limitations on precision?

Aside from these improvements, I don't think that this is a high-priority issue now that @BFCombes has guidance on how to move forward.
